src/main/cpp/util/file_windows.cc

```diff
@@ -549,6 +549,99 @@ bool JunctionResolver::Resolve(const WCHAR* path, unique_ptr<WCHAR[]>* result) {
   return Resolve(path, result, kMaximumJunctionDepth);
 }
 
+class SymlinkResolver {
+ public:
+  SymlinkResolver();
+
+  // Resolves symlink to its actual path.
+  //
+  // Returns true if `path` is not a symlink and it exists.
+  // Returns true if `path` is a symlink and can be successfully resolved.
+  // Returns false otherwise.
+  //
+  // If `result` is not nullptr and the method returned true, then this will be
+  // reset to point to a new WCHAR buffer containing the resolved path.
+  // If `path` is a symlink, this will be the resolved path, otherwise
+  // it will be a copy of `path`.
+  bool Resolve(const WCHAR* path, std::unique_ptr<WCHAR[]>* result);
+
+ private:
+  // Symbolic Link Reparse Data Buffer is described at:
+  // https://msdn.microsoft.com/en-us/library/cc232006.aspx
+  typedef struct _ReparseSymbolicLinkData {
+    static const int kSize = MAXIMUM_REPARSE_DATA_BUFFER_SIZE;
+    ULONG ReparseTag;
+    USHORT ReparseDataLength;
+    USHORT Reserved;
+    USHORT SubstituteNameOffset;
+    USHORT SubstituteNameLength;
+    USHORT PrintNameOffset;
+    USHORT PrintNameLength;
+    ULONG Flags;
+    WCHAR PathBuffer[1];
+  } ReparseSymbolicLinkData;
+
+  uint8_t reparse_buffer_bytes_[ReparseSymbolicLinkData::kSize];
+  ReparseSymbolicLinkData* reparse_buffer_;
+};
+
+SymlinkResolver::SymlinkResolver()
+    : reparse_buffer_(
+          reinterpret_cast<ReparseSymbolicLinkData*>(reparse_buffer_bytes_)) {
+}
+
+bool SymlinkResolver::Resolve(const WCHAR* path, unique_ptr<WCHAR[]>* result) {
+  AutoHandle handle(CreateFileW(path,
+                                FILE_READ_EA,
+                                FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                                NULL,
+                                OPEN_EXISTING,
+                                FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OPEN_REPARSE_POINT,
+                                NULL));
+  if (!handle.IsValid()) {
```

Comment:
In my experience, CreateFileW(..., FILE_ATTRIBUTE_NORMAL | ...) fails for directories (and junctions, which are also directories), therefore this failure may just mean that "path" refers to a directory.

Reply:
Then I think we can call `GetFileAttributesW` first to check if the `path` refers to a reparse point or not. If it does, we can just use `FILE_FLAG_OPEN_REPARSE_POINT`, otherwise, we skip and return the original path. WDYT?

Reply:
And we can tell from the attribute if the `path` is a directory or not, if it is, we set `FILE_FLAG_BACKUP_SEMANTICS`

Reply:
Yes, I think calling GetFileAttributeW then CreateFileW (with FILE_ATTRIBUTE_NORMAL or FILE_FLAG_BACKUP_SEMANTICS, depending on the attributes) is as good as calling CreateFileW first with FILE_FLAG_BACKUP_SEMANTICS then with FILE_ATTRIBUTE_NORMAL -- both approaches achieve the same goal and both are susceptible to file system races (i.e. when the file is modified between the two function calls).

Reply:
OK, it's done. But I think we don't need `FILE_ATTRIBUTE_NORMAL` for file symlink because in the document it says `This attribute is valid only if used alone.`
https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-createfilea